model/riscv_insts_zicsr.sail

```diff
@@ -109,27 +109,13 @@ function readCSR csr : csreg -> xlenbits = {
     (0x343,  _) => mtval,
     (0x344,  _) => mip.bits(),
 
-    (0x3A0,  _) => pmpReadCfgReg(0),           // pmpcfg0
-    (0x3A1, 32) => pmpReadCfgReg(1),           // pmpcfg1
-    (0x3A2,  _) => pmpReadCfgReg(2),           // pmpcfg2
-    (0x3A3, 32) => pmpReadCfgReg(3),           // pmpcfg3
-
-    (0x3B0,  _) => pmpaddr0,
-    (0x3B1,  _) => pmpaddr1,
-    (0x3B2,  _) => pmpaddr2,
-    (0x3B3,  _) => pmpaddr3,
-    (0x3B4,  _) => pmpaddr4,
-    (0x3B5,  _) => pmpaddr5,
-    (0x3B6,  _) => pmpaddr6,
-    (0x3B7,  _) => pmpaddr7,
-    (0x3B8,  _) => pmpaddr8,
-    (0x3B9,  _) => pmpaddr9,
-    (0x3BA,  _) => pmpaddr10,
-    (0x3BB,  _) => pmpaddr11,
-    (0x3BC,  _) => pmpaddr12,
-    (0x3BD,  _) => pmpaddr13,
-    (0x3BE,  _) => pmpaddr14,
-    (0x3BF,  _) => pmpaddr15,
+    // pmpcfgN
+    (0x3A @ idx : bits(4),  _) if idx[0] == bitzero | sizeof(xlen) == 32 => pmpReadCfgReg(unsigned(idx)),
```

Comment:
would it be easier to define idx to be 6 bits, and >>2 for cfg case, and leave it alone for the address case?

Reply:
Ah, OK, then why not (idx+16)>>2? (+ worrying about overflow...)

Reply:
Not sure I follow. Can you post some code to explain?

Reply:
(0b001111 @ (idx : bits(6))+0x10,  _) => pmpWriteAddrReg(idx, value);
so what was 
originally matching   0011 1011 ???? 
      now is matching  0011 1100 ????
originally matching   0011 1100 ???? 
      now is matching  0011 1101 ???? etc
which generalizes to 0011 11?? ????
 and you can just pick off the 6 LSBs as the index number.

For config, you need to take the 6bit idx and >>2, using the remaining bits as index

Reply:
That isn't valid Sail code though. If you can find something that's simpler and compiles I'm all for it!

Reply:
I'm pretty sure it isn't; I don't know Sail enough to make it value. I suspect you need to explicit convert idx:bits(6) to  a 6 bit NAT, add 16 to it, and convert it back to bits(6) to get it to work. Other than that: why won't it compile?

Reply:
Ah I see what you are proposing now. Unfortunately it won't work because:

* Sail doesn't support refinement types for `bits()`, so `bits(6) + 0x10` isn't a valid type.
* Addition on `bits()` wraps anyway. It doesn't overflow past the `@` which is just for concatenation.

Given that Sail doesn't support that, this is the best available option I know of. I don't think it's too bad really.

Reply:
I'm nearly following the syntax, but that sounds more like what I had in mind - that's matching on a range check
Splitting it into bits(5)@bits(7) as opposed to bits(8)@bits(4) makes it look more understandable to me:
 0x07 @  idx: bits(7) if 0x30 <=_u idx <_u 0x70 => pmpReadAddrReg(unsigned((idx - 0x30)))

Reply:
Unfortunately it doesn't seem like the compiler is quite smart enough to figure things out with Alasdair's suggestion:

    (x : bits(8) @ idx : bits(4), _) if (0x3B <=_u x) & (x <=_u 0x3E) => pmpReadAddrReg(unsigned((x - 0x3B) @ idx)),

I get:

```
Type error:
model/riscv_insts_zicsr.sail:116.73-115:
116 |    (x : bits(8) @ idx : bits(4), _) if (0x3B <=_u x) & (x <=_u 0x3E) => pmpReadAddrReg(unsigned((x - 0x3B) @ idx)),
    |                                                                         ^----------------------------------------^
    | Could not resolve quantifiers for pmpReadAddrReg
    | * (0 <= 'ex9543# & 'ex9543# <= 63)
```

I also tried this but it had the same error:

    (idx : bits(12),  _) if (0x3B0 <=_u idx) & (idx <_u 0x3F0) => pmpReadAddrReg(unsigned(idx - 0x3B0)),

And this:

    (idx : bits(12),  _) if (0x3B0 <=_u idx) & (idx <_u 0x3F0) => pmpReadAddrReg(unsigned(idx) - 944),

I think the original code is ok. Let's leave this and if someone can up with an improvement later then they can make a PR.

Reply:
OK, then, I think this is good enough to merge.